src/seq_mv/csr_matop_device.c

```diff
@@ -2165,57 +2236,95 @@ hypre_CSRMatrixTriLowerUpperSolveCusparse(char             uplo,
       return hypre_error_flag;
    }
 
-   cusparseHandle_t handle = hypre_HandleCusparseHandle(hypre_handle());
-   cusparseMatDescr_t descr = hypre_CSRMatrixGPUMatDescr(A);
+#if CUSPARSE_VERSION < CUSPARSE_SPSV_VERSION
+   /* cusparse (legacy API) requires sorted rows. Sort and save in CSR's (sj, sa) */
+   hypre_CSRMatrixSortRowOutOfPlace(A);
+#endif
 
-   if ( !A_sj && !A_sa )
+   /* setup csrsvdata in CSR: modify the diagonal (once) */
+   if (!hypre_CSRMatrixCsrsvData(A))
    {
-      hypre_CSRMatrixSortedJ(A) = A_sj = hypre_TAlloc(HYPRE_Int, nnzA, HYPRE_MEMORY_DEVICE);
-      hypre_CSRMatrixSortedData(A) = A_sa = hypre_TAlloc(HYPRE_Complex, nnzA, HYPRE_MEMORY_DEVICE);
-      hypre_TMemcpy(A_sj, A_j, HYPRE_Int, nnzA, HYPRE_MEMORY_DEVICE, HYPRE_MEMORY_DEVICE);
-      hypre_TMemcpy(A_sa, A_a, HYPRE_Complex, nnzA, HYPRE_MEMORY_DEVICE, HYPRE_MEMORY_DEVICE);
+      hypre_CSRMatrixCsrsvData(A) = hypre_CsrsvDataCreate();
+      hypre_CsrsvData *csrsv_data = hypre_CSRMatrixCsrsvData(A);
+
+      hypre_CsrsvDataMatData(csrsv_data) = hypre_TAlloc(HYPRE_Complex, nnzA, HYPRE_MEMORY_DEVICE);
+
+      hypre_CSRMatrixData(A) = hypre_CsrsvDataMatData(csrsv_data);
+#if CUSPARSE_VERSION >= CUSPARSE_SPSV_VERSION
+      hypre_TMemcpy(hypre_CsrsvDataMatData(csrsv_data), A_a, HYPRE_Complex, nnzA,
+                    HYPRE_MEMORY_DEVICE, HYPRE_MEMORY_DEVICE);
+#else
+      hypre_TMemcpy(hypre_CsrsvDataMatData(csrsv_data), hypre_CSRMatrixSortedData(A), HYPRE_Complex, nnzA,
+                    HYPRE_MEMORY_DEVICE, HYPRE_MEMORY_DEVICE);
+      hypre_CSRMatrixJ(A) = hypre_CSRMatrixSortedJ(A);
+#endif
+
+      /* if (l1_norms), replace A's diag with l1_norm, and
+       * replace zero diag with inf. so as to skip relaxation for this unknown */
+      err = hypre_CSRMatrixReplaceDiagDevice(A, l1_norms, INFINITY, 0.0);
```

Comment:
Minor observation: we don't need to return `err`, just checking `hypre_error_flag` in the next `if` is enough since it's a global variable

Reply:
Do you mean the return in `hypre_CSRMatrixReplaceDiagDevice`? In debug mode, it returns the number of "structural" zero in the diagonal of a matrix. Not sure what I should change. Thanks.

Reply:
Ok! I assumed it was returning `hypre_error_flag` as we normally do in other `hypre_` functions, but I didn't double check that was the case, and it wasn't. Would it be more correct to check for the structural zero and give the `hypre_error_w_msg` from inside `hypre_CSRMatrixReplaceDiagDevice`? That way, we could make it return `hypre_error_flag`. My understanding is that returning `hypre_error_flag` is the preferred behavior in hypre, please correct me if I'm wrong

Reply:
Sounds good. Thanks!

Reply:
Does it look good to you now?

Reply:
This looks great! Thanks for doing the changes